Create new UnionFind struct for DBSCAN

aprokop opened the pull request:
This makes the code more readable, and prepares a stage for the dense
box implementation of the algorithm.

dalg24 commented on src/details/ArborX_DetailsDBSCANCallback.hpp:
```diff
@@ -123,7 +122,10 @@ struct DBSCANCallback
+  void union_1way(int i, int j) const { labels_(i) = representative(j); }
+
+  KOKKOS_FUNCTION
+  void union_2way(int i, int j) const
```
Why union "one" and "two" way?  I don't get it.

aprokop replied:
Union-find algorithm essentially merges two (implicit) trees. In general, it does not matter which one is a root, and which one is merged into, and is usually based on which representative is smaller. However, for the boundary points, it matters, as otherwise it may lead to the bridging effect. So the boundary point tree has to be merged into the core point tree, which is what I call `union_1way`, which simply assigns a label.

dalg24 replied:
I still don't get the name

aprokop replied:
Given what I described, how would you propose to call them?

masterleinad replied:
Maybe be more explicit via `union_boundary_core` or such?

aprokop replied:
Union-find structure is a general concept outside of DBSCAN. I would want to limit its scope to its basic functionality, including the names.

masterleinad replied:
I would just not include `union_1way` in `UnionFind` then (and have the behavior in-line as before) and don't change the name of `union_2way`.

aprokop replied:
@masterleinad I need union 1 way in UnionFind, as I do not want to expose labels.

@masterleinad @dalg24 Would you guys feel better if I change the signature to
`union(int i, int j, bool one_way = false);`? It may be slightly less performant, as it will introduce an extra unnecessary `if` check, but should not be a biggie.

masterleinad replied:
I think it's fine to have two member functions in that case but they should be self-describing. What you essentially want is a version that assigns that updates the left tree ( or the right tree), isn't it?
 If so, `union_merge_left` and `union_merge_any` (or just `union`) would be fine with me. Alternatively, introducing a template argument `force_merge_left` should not have any runtime overhead and might be nicer by providing only one function.

aprokop replied:
> What you essentially want is a version that assigns that updates the left tree ( or the right tree), isn't it?

Well, similar, but there is no left tree and right tree. There are two trees, one consisting of a single vertex (corresponding to the boundary point), and one other tree. To me, `left` is completely confusing, and does not make sense.

masterleinad replied:
OK. Then calling this function has special requisites, i.e. one of them having only a single vertex, in that case I would certainly use two different names, maybe `union_with_single_vertex_tree` or so.

aprokop commented on src/ArborX_DBSCAN.hpp:
```diff
@@ -136,6 +136,11 @@ dbscan(ExecutionSpace const &exec_space, Primitives const &primitives,
+  Kokkos::View<int *, MemorySpace> labels(
+      Kokkos::ViewAllocateWithoutInitializing("ArborX::DBSCAN::labels"), n);
+  ArborX::iota(exec_space, labels);
+  Details::UnionFind<MemorySpace> union_find{labels};
```
Yeah, I guess I don't have to do it here. I took this patch from the dense box branch, it makes more sense there. I'll put it back.

dalg24 commented on src/ArborX_DBSCAN.hpp:
```diff
@@ -149,15 +149,16 @@ dbscan(ExecutionSpace const &exec_space, Primitives const &primitives,
   Kokkos::View<int *, MemorySpace> labels(
       Kokkos::ViewAllocateWithoutInitializing("ArborX::DBSCAN::labels"), n);
   ArborX::iota(exec_space, labels);
+  Details::UnionFind<MemorySpace> union_find{labels};
```
Why do you want to expose this?  This is an implementation of the callback.  Take the labels like you did before and create your union find in the callback constructor.

aprokop replied:
Well, I could do that in this PR, but I would have to move it out in the dense box PR, as I will also need to update this outside of the callback. So, rather than passing the labels around I like it better to pass union-find struct around. This also makes sense logically, as we could change the way union-find struct implemented (if desired) without changing the callback.

aprokop commented:
@dalg24 Do my comments address your concerns?

aprokop commented:
Two SYCL timeouts, but otherwise tests pass.

aprokop commented:
@dalg24 Should have addressed your comments.

I renamed `union_2way` and `union_1way` to `merge` and `merge_into`. Could not use `union` because it's a reserved word. Union-find is also called merge-find sometimes, so these names make some sense. I could have renamed `UnionFind` as a `MergeFind`, but I like `UnionFind` better.

masterleinad reviewed (approved):
Fine with me.

dalg24 reviewed (approved):
I'd prefer if you squash merged or rewrote history
